File: ios/LiveActivity/HomeWidgetBridge.swift
import Flutter
import Foundation

#if canImport(WidgetKit)
  import WidgetKit
#endif

/// MethodChannel between the Flutter app and the Home Screen widgets.
///
/// The app is the only writer. Everything arriving here is already a finished
/// JSON snapshot — this class validates it, stores it in the App Group and asks
/// WidgetKit to re-render.
final class HomeWidgetBridge {
  static let channelName = "trainlibre.widgets/home_screen"

  func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "isSupported":
      result(Self.isSupported)
    case "writeSnapshot":
      writeSnapshot(args: call.arguments as? [String: Any], result: result)
    case "writeSharedFile":
      writeSharedFile(args: call.arguments as? [String: Any], result: result)
    case "sharedFileExists":
      let name = (call.arguments as? [String: Any])?["name"] as? String ?? ""
      guard let url = TrainLibreHomeWidget.sharedFileURL(named: name) else {
        result(false)
        return
      }
      result(FileManager.default.fileExists(atPath: url.path))
    case "clearSnapshot":
      TrainLibreHomeWidget.defaults?.removeObject(forKey: TrainLibreHomeWidget.snapshotKey)
      Self.reloadTimelines()
      result(true)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  /// The widgets are gated to iOS 18 in the bundle, so anything older can skip
  /// building snapshots altogether rather than write into a void.
  private static var isSupported: Bool {
    #if canImport(WidgetKit)
      if #available(iOS 18.0, *) { return true }
    #endif
    return false
  }

  private func writeSnapshot(args: [String: Any]?, result: @escaping FlutterResult) {
    guard
      let json = args?["json"] as? String,
      let data = json.data(using: .utf8),
      // Decode before storing: a snapshot the widget cannot read is worse than
      // no snapshot, because it would silently freeze the last good render.
      (try? JSONDecoder().decode(HomeWidgetSnapshot.self, from: data)) != nil,
      let defaults = TrainLibreHomeWidget.defaults
    else {
      result(false)
      return
    }

    defaults.set(json, forKey: TrainLibreHomeWidget.snapshotKey)
    Self.reloadTimelines()
    result(true)
  }

  /// Stores a binary asset — today only the muscle heatmap — in the App Group
  /// container, where the widget extension can read it as a file.
  ///
  /// Kept out of `UserDefaults`: a PNG is far past what a plist should carry,
  /// and the widget would have to decode it on every timeline render.
  private func writeSharedFile(args: [String: Any]?, result: @escaping FlutterResult) {
    guard
      let name = args?["name"] as? String,
      let url = TrainLibreHomeWidget.sharedFileURL(named: name)
    else {
      result(false)
      return
    }

    // A nil payload is a delete — the app uses it when a workout is removed and
    // its heatmap must not outlive it.
    guard let bytes = args?["bytes"] as? FlutterStandardTypedData else {
      try? FileManager.default.removeItem(at: url)
      Self.reloadTimelines()
      result(true)
      return
    }

    do {
      try bytes.data.write(to: url, options: .atomic)
      Self.sweepSiblings(of: url)
      Self.reloadTimelines()
      result(true)
    } catch {
      result(false)
    }
  }

  /// Removes older files sharing the new file's prefix.
  ///
  /// Heatmaps are named per workout so a snapshot can never point at the wrong
  /// session's map. The cost of that is that every finished workout would leave
  /// a file behind, so the newest write clears the ones it superseded.
  private static func sweepSiblings(of url: URL) {
    let name = url.lastPathComponent
    guard
      let separator = name.range(of: "_", options: .backwards),
      case let prefix = String(name[name.startIndex..<separator.upperBound]),
      !prefix.isEmpty,
      let directory = TrainLibreHomeWidget.containerURL,
      let entries = try? FileManager.default.contentsOfDirectory(
        at: directory,
        includingPropertiesForKeys: nil
      )
    else { return }

    for entry in entries
    where entry.lastPathComponent.hasPrefix(prefix) && entry.lastPathComponent != name {
      try? FileManager.default.removeItem(at: entry)
    }
  }

  private static func reloadTimelines() {
    #if canImport(WidgetKit)
      if #available(iOS 18.0, *) {
        for kind in TrainLibreHomeWidget.allKinds {
          WidgetCenter.shared.reloadTimelines(ofKind: kind)
        }
      }
    #endif
  }
}

File: android/app/src/main/kotlin/com/rfivesix/trainlibre/widgets/HomeWidgetBridge.kt
package com.rfivesix.trainlibre.widgets

import android.content.Context
import com.rfivesix.trainlibre.widgets.snapshot.HomeWidgetStore
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * MethodChannel between the Flutter app and the home screen widgets.
 *
 * The app is the only writer. Everything arriving here is already a finished
 * JSON snapshot — this class validates it, stores it and asks the widgets to
 * re-render. Counterpart to `ios/LiveActivity/HomeWidgetBridge.swift`, and it
 * answers the exact same five methods.
 */
class HomeWidgetBridge(private val context: Context) {

    companion object {
        const val CHANNEL_NAME = "trainlibre.widgets/home_screen"
    }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    fun handle(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            // Unlike iOS, where the widgets are gated to iOS 18, there is no
            // version below which this does nothing — Glance covers the app's
            // whole minSdk range.
            "isSupported" -> result.success(true)
            "writeSnapshot" -> writeSnapshot(call, result)
            "writeSharedFile" -> writeSharedFile(call, result)
            "sharedFileExists" -> {
                val name = call.argument<String>("name").orEmpty()
                result.success(HomeWidgetStore.sharedFileExists(context, name))
            }
            "clearSnapshot" -> {
                val cleared = HomeWidgetStore.clearSnapshot(context)
                refreshThen(result, cleared)
            }
            else -> result.notImplemented()
        }
    }

    private fun writeSnapshot(call: MethodCall, result: MethodChannel.Result) {
        val json = call.argument<String>("json")
        if (json == null) {
            result.success(false)
            return
        }
        // The store decodes before it stores: a snapshot the widget cannot read
        // is worse than no snapshot, because it would silently freeze the last
        // good render.
        val written = HomeWidgetStore.writeSnapshot(context, json)
        if (!written) {
            result.success(false)
            return
        }
        refreshThen(result, true)
    }

    /**
     * Stores a binary asset — today only the muscle heatmap.
     *
     * A null payload is a delete: the app uses it when a workout is removed and
     * its heatmap must not outlive it.
     */
    private fun writeSharedFile(call: MethodCall, result: MethodChannel.Result) {
        val name = call.argument<String>("name")
        if (name == null) {
            result.success(false)
            return
        }
        val bytes = call.argument<ByteArray>("bytes")
        val ok = if (bytes == null) {
            HomeWidgetStore.deleteSharedFile(context, name)
        } else {
            HomeWidgetStore.writeSharedFile(context, name, bytes)
        }
        if (!ok) {
            result.success(false)
            return
        }
        refreshThen(result, true)
    }

    /**
     * Redraws every widget, then answers the app.
     *
     * The order matters more than it looks. The app's most important write is
     * the one it makes while being backgrounded — that is the moment just before
     * the home screen becomes visible. Answering first and redrawing afterwards
     * loses exactly that update: Android freezes a backgrounded process within
     * moments, and a coroutine still on its way to `updateAll` goes with it.
     * Holding the channel result open keeps the app alive until the widgets have
     * actually been redrawn.
     *
     * Every widget is redrawn rather than guessing which section changed — a
     * snapshot write is already debounced to one per user action, so the saving
     * would buy nothing and the guess could be wrong.
     */
    private fun refreshThen(result: MethodChannel.Result, value: Boolean) {
        scope.launch {
            HomeWidgetRefresher.refreshAll(context)
            withContext(Dispatchers.Main) { result.success(value) }
        }
    }
}
